File: packages/sources/sdk-ts/src/learning/exporter.ts
import { LearningSample } from './types';
import { validateLearningSample } from './validator';
import { createHash } from 'crypto';

// Derived from Truth Sources
// schemas/v2/learning/mplp-learning-sample-core.schema.json
// schemas/v2/taxonomy/learning-taxonomy.yaml
const PROTOCOL_VERSION = "1.0.0";
const GOVERNANCE = "MPGC";
const FROZEN = true;
const FREEZE_DATE = "2025-12-03";
const LEARNING_TAXONOMY_SHA256 = "6dbc28bb2dd1fa3a5e1ec51c20a0cc94885cce22989c0f9eb550914482442464e";

export type ExportMode = "fail_fast" | "skip_invalid";

export interface ExportOptions {
    mode?: ExportMode;
}

export interface LearningManifestLite {
    protocol_version: string;
    governance: string;
    frozen: boolean;
    freeze_date: string;
    exported_at: string;
    count: number;
    invalid_count: number;
    ndjson_sha256: string;
    learning_taxonomy_sha256: string;
    schema_bundle_sha256: string | null;
    schema_bundle_hash_method: "sha256" | "none";
    notes?: string[];
}

export function readTruthMeta() {
    return {
        protocol_version: PROTOCOL_VERSION,
        governance: GOVERNANCE,
        frozen: FROZEN,
        freeze_date: FREEZE_DATE,
        schema_bundle_sha256: null,
        schema_bundle_hash_method: "none" as const,
        learning_taxonomy_sha256: LEARNING_TAXONOMY_SHA256,
    };
}

export function canonicalJsonLine(obj: unknown): string {
    // Recursive sort keys
    const sortKeys = (o: any): any => {
        if (Array.isArray(o)) {
            return o.map(sortKeys);
        } else if (o !== null && typeof o === 'object') {
            return Object.keys(o).sort().reduce((acc, key) => {
                acc[key] = sortKeys(o[key]);
                return acc;
            }, {} as any);
        }
        return o;
    };
    return JSON.stringify(sortKeys(obj));
}

export function sha256Hex(input: Uint8Array | string): string {
    const hash = createHash('sha256');
    hash.update(input);
    return hash.digest('hex');
}

export function validateExportSamples(samples: LearningSample[], options?: ExportOptions): { valid: LearningSample[]; invalid: Array<{ index: number; error: unknown }>; } {
    const valid: LearningSample[] = [];
    const invalid: Array<{ index: number; error: unknown }> = [];
    const mode = options?.mode || "fail_fast";

    for (let i = 0; i < samples.length; i++) {
        const sample = samples[i];
        const result = validateLearningSample(sample);
        if (result.valid) {
            valid.push(sample);
        } else {
            invalid.push({ index: i, error: result.errors });
            if (mode === "fail_fast") {
                // Stop immediately
                break;
            }
        }
    }
    return { valid, invalid };
}

export function exportLearningNdjson(samples: LearningSample[], options?: ExportOptions): string {
    const { valid, invalid } = validateExportSamples(samples, options);
    const mode = options?.mode || "fail_fast";

    if (mode === "fail_fast" && invalid.length > 0) {
        const firstError = invalid[0];
        const error: any = new Error(`Validation failed for sample at index ${firstError.index}`);
        error.code = "LEARNING_EXPORT_VALIDATION_FAILED";
        error.invalid_count = invalid.length;
        error.first_error_index = firstError.index;
        error.details = firstError.error;
        throw error;
    }

    if (valid.length === 0) {
        return "";
    }

    return valid.map(s => canonicalJsonLine(s)).join('\n') + '\n';
}

export function generateLearningManifestLite(args: { samples: LearningSample[]; ndjson: string; options?: ExportOptions }): LearningManifestLite {
    const { samples, ndjson, options } = args;
    const validation = validateExportSamples(samples, options);

    return {
        ...readTruthMeta(),
        exported_at: new Date().toISOString(),
        count: validation.valid.length,
        invalid_count: validation.invalid.length,
        ndjson_sha256: sha256Hex(ndjson),
        notes: []
    };
}

File: packages/sources/sdk-py/src/mplp/learning/exporter.py
import json
import hashlib
import os
from typing import List, Optional, Dict, Any, Literal, Tuple
from datetime import datetime, timezone
from pydantic import BaseModel
from .types import LearningSample
from .validator import validate_learning_sample

ExportMode = Literal["fail_fast", "skip_invalid"]

class ExportOptions(BaseModel):
    mode: ExportMode = "fail_fast"

class LearningManifestLite(BaseModel):
    protocol_version: str
    governance: str
    frozen: bool
    freeze_date: str
    exported_at: str
    count: int
    invalid_count: int
    ndjson_sha256: str
    learning_taxonomy_sha256: str
    schema_bundle_sha256: Optional[str]
    schema_bundle_hash_method: Literal["sha256", "none"]
    notes: Optional[List[str]] = []

def read_truth_meta() -> Dict[str, Any]:
    # Load from truth_snapshot.json in the same directory
    current_dir = os.path.dirname(__file__)
    snapshot_path = os.path.join(current_dir, "truth_snapshot.json")
    
    with open(snapshot_path, 'r', encoding='utf-8') as f:
        snapshot = json.load(f)
        
    return {
        "protocol_version": snapshot["protocol_version"],
        "governance": snapshot["governance"],
        "frozen": snapshot["frozen"],
        "freeze_date": snapshot["freeze_date"],
        "schema_bundle_sha256": snapshot["schema_bundle_sha256"],
        "schema_bundle_hash_method": snapshot["schema_bundle_hash_method"],
        "learning_taxonomy_sha256": snapshot["learning_taxonomy_sha256"],
    }

def canonical_json_line(obj: Any) -> str:
    # Pydantic models to dict first
    if isinstance(obj, BaseModel):
        data = obj.model_dump(mode='json') # handles datetime serialization etc
    else:
        data = obj
    
    # Sort keys and compact separators
    return json.dumps(data, separators=(',', ':'), sort_keys=True)

def sha256_hex(data: bytes | str) -> str:
    if isinstance(data, str):
        data = data.encode('utf-8')
    return hashlib.sha256(data).hexdigest()

def validate_export_samples(samples: List[LearningSample], options: Optional[ExportOptions] = None) -> Tuple[List[LearningSample], List[Tuple[int, Any]]]:
    valid = []
    invalid = []
    mode = options.mode if options else "fail_fast"

    for i, sample in enumerate(samples):
        result = validate_learning_sample(sample)
        if result.valid:
            valid.append(sample)
        else:
            invalid.append((i, result.errors))
            if mode == "fail_fast":
                break
    return valid, invalid

class LearningExportValidationError(Exception):
    def __init__(self, message, code, invalid_count, first_error_index, details):
        super().__init__(message)
        self.code = code
        self.invalid_count = invalid_count
        self.first_error_index = first_error_index
        self.details = details

def export_learning_ndjson(samples: List[LearningSample], options: Optional[ExportOptions] = None) -> str:
    valid, invalid = validate_export_samples(samples, options)
    mode = options.mode if options else "fail_fast"

    if mode == "fail_fast" and invalid:
        first_error = invalid[0]
        raise LearningExportValidationError(
            message=f"Validation failed for sample at index {first_error[0]}",
            code="LEARNING_EXPORT_VALIDATION_FAILED",
            invalid_count=len(invalid),
            first_error_index=first_error[0],
            details=first_error[1]
        )

    if not valid:
        return ""

    lines = [canonical_json_line(s) for s in valid]
    return "\n".join(lines) + "\n"

def generate_learning_manifest_lite(samples: List[LearningSample], ndjson: str, options: Optional[ExportOptions] = None) -> LearningManifestLite:
    # Re-validate to get counts (or trust caller, but we need invalid_count)
    valid, invalid = validate_export_samples(samples, options)
    
    meta = read_truth_meta()
    
    return LearningManifestLite(
        **meta,
        exported_at=datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z'), # Ensure ISO format
        count=len(valid),
        invalid_count=len(invalid),
        ndjson_sha256=sha256_hex(ndjson),
        notes=[]
    )
